wgpu-core/src/device/queue.rs

```diff
@@ -191,7 +192,7 @@ impl<G: GlobalIdentityHandlerFactory> Global<G> {
         let device = device_guard
             .get_mut(queue_id)
             .map_err(|_| DeviceError::Invalid)?;
-        let (buffer_guard, _) = hub.buffers.read(&mut token);
+        let (mut buffer_guard, _) = hub.buffers.write(&mut token);
```

Comment:
It's really unfortunate to lock it here for writing. Our locking is imperfect, and we should try to steer away from making it kill our perf.
Here, how do you feel about the following: we do everything as usual read-only, but in the end, if we see that some ranges need to be filled, we `drop(buffer_guard)` and then obtain a new writing guard, to deal with the init ranges specifically?